### PR TITLE
Parser of MSBuild binlog query language 

### PR DESCRIPTION
**Pull request by yuchong-pan**
~~_Note: This PR should be reviewed after #79 has been merged._~~
~~_This PR is still a work in progress._~~

Grammer supported:
```
<SLASHES>      ::= "/"
                 | "//"
<MSG_NODE>     ::= <SLASHES> "message"
                 | <SLASHES> "warning"
                 | <SLASHES> "error"
<TASK_NODE>    ::= "/" "task"
                 | "/" "task" <MSG_NDOE>
<TARGET_NODE>  ::= "/" "target"
                 | "/" "target" <TASK_NODE>
                 | "/" "target" <MSG_NODE>
<PROJECT_NODE> ::= "/" "project"
                 | "/" "project" <TARGET_NODE>
                 | "/" "project" <TASK_NODE>
                 | "/" "project" <MSG_NODE>
<QUERY>        ::= <PROJECT_NODE>
                 | <TARGET_NODE>
                 | <TASK_NODE>
                 | <MSG_NODE>
```

Refactored LL(1) grammar:
```
<SLASHES>      ::= "/"
                 | "//"
<TEMP1>        ::= "message"
                 | "warning"
                 | "error"
<MSG_NODE>     ::= <SLASHES> <TEMP1>
<TEMP2>        ::= empty
                 | <MSG_NODE>
<TEMP4>        ::= <TEMP1>
                 | "task" <TEMP2>
<TEMP5>        ::= empty
                 | "/" <TEMP4>
                 | "//" <TEMP1>
<TEMP7>        ::= <TEMP1>
                 | "target" <TEMP5>
                 | "task" <TEMP2>
<TEMP8>        ::= empty
                 | "/" <TEMP7>
                 | "//" <TEMP1>
<TEMP9>        ::= "project" <TEMP8>
                 | "target" <TEMP5>
                 | "task" <TEMP2>
                 | <TEMP1>
<QUERY>        ::= "/" <TEMP9>
                 | "//" <TEMP1>
```

### CODE REVIEW

File: src/MSBuildBinLogQuery/Parse/Parser.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseQuery()
+        {
+            if (_scanner.Token is SingleSlashToken)
+            {
+                Consume<SingleSlashToken>();
+                return ParseProjectNodeOrTargetNodeOrTaskNodeOrSingleSlashLogNode();
```
Comment (rainersigwald) [inline src/MSBuildBinLogQuery/Parse/Parser.cs]: These names are a real mouthful. Is there one that's more meaningful in the domain of the MSBuild graph?

File: src/MSBuildBinLogQuery/Parse/Parser.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                var next = ParseLogNodeOrNull();
+                return new TaskNode(next);
+            }
+
+            return ParseLogNodeWithType(LogNodeType.Direct);
+        }
+
+        private AstNode ParseTargetNodeOrTaskNodeOrLogNodeOrNull()
+        {
+            if (_scanner.Token is SingleSlashToken)
```
Comment (rainersigwald) [inline src/MSBuildBinLogQuery/Parse/Parser.cs]: personally I'd find pattern matching clearer here, but if you prefer these go for it

File: src/MSBuildBinLogQuery/Parse/Parser.cs (hunk 1 above)
Comment (yuchong-pan) [reply src/MSBuildBinLogQuery/Parse/Parser.cs]: Not sure for this one. But for others can we say `ParseNodeUnderProject` for `ParseTargetNodeOrTaskNodeOr...`

File: src/MSBuildBinLogQuery/Parse/Parser.cs (hunk 2 above)
Comment (yuchong-pan) [reply src/MSBuildBinLogQuery/Parse/Parser.cs]: What do you mean by "pattern matching" here (I think it has different meanings in different context..)? Could you give an example?

Comment (rainersigwald) [reply src/MSBuildBinLogQuery/Parse/Parser.cs]: I'm talking about [C# 7 `switch` type-based pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch#type-pattern). This method, for example, could be expressed as:

```C#
        private AstNode ParseTargetNodeOrTaskNodeOrLogNodeOrNull()
        {
            switch (_scanner.Token)
            {
                case SingleSlashToken _:
                    Consume<SingleSlashToken>();
                    return ParseTargetNodeOrTaskNodeOrSingleSlashLogNode();
                case DoubleSlashToken _:
                    Consume<DoubleSlashToken>();
                    return ParseLogNodeWithType(LogNodeType.All);
            }

            return null;
        }
```

Comment (yuchong-pan) [reply src/MSBuildBinLogQuery/Parse/Parser.cs]: I think It's better than what I have. Thanks!

Comment (yuchong-pan) [reply src/MSBuildBinLogQuery/Parse/Parser.cs]: Done.

File: src/MSBuildBinLogQuery/Parse/Parser.cs (hunk 1 above)
Comment (yuchong-pan) [reply src/MSBuildBinLogQuery/Parse/Parser.cs]: Done.

**Pull request merged**